include/swift/AST/DiagnosticEngine.h

```diff
@@ -425,25 +439,89 @@ namespace swift {
     InFlightDiagnostic &highlightChars(SourceLoc Start, SourceLoc End);
 
+    static const char *fixItStringFor(const FixItID id);
+
+    /// Add a token-based replacement fix-it to the currently-active
+    /// diagnostic.
+    template <typename... ArgTypes>
+    InFlightDiagnostic &
+    fixItReplace(SourceRange R, StructuredFixIt<ArgTypes...> fixIt,
+                 typename detail::PassArgument<ArgTypes>::type... VArgs) {
+      SmallVector<DiagnosticArgument, 3> Args;
+      DiagnosticArgument DiagArgs[] = {
+        DiagnosticArgument(0), std::move(VArgs)...
+      };
+      Args.append(DiagArgs + 1, DiagArgs + 1 + sizeof...(VArgs));
+      return fixItReplace(R, fixItStringFor(fixIt.ID), std::move(Args));
+    }
+
+    /// Add a character-based replacement fix-it to the currently-active
+    /// diagnostic.
+    template <typename... ArgTypes>
+    InFlightDiagnostic &
+    fixItReplaceChars(SourceLoc Start, SourceLoc End,
+                      StructuredFixIt<ArgTypes...> fixIt,
+                      typename detail::PassArgument<ArgTypes>::type... VArgs) {
+      SmallVector<DiagnosticArgument, 3> Args;
+      DiagnosticArgument DiagArgs[] = {
+        DiagnosticArgument(0), std::move(VArgs)...
+      };
+      Args.append(DiagArgs + 1, DiagArgs + 1 + sizeof...(VArgs));
+      return fixItReplaceChars(Start, End, fixItStringFor(fixIt.ID),
+                               std::move(Args));
+    }
+
+    /// Add an insertion fix-it to the currently-active diagnostic.
+    template <typename... ArgTypes>
+    InFlightDiagnostic &
+    fixItInsert(SourceLoc L, StructuredFixIt<ArgTypes...> fixIt,
+                typename detail::PassArgument<ArgTypes>::type... VArgs) {
+      SmallVector<DiagnosticArgument, 3> Args;
+      DiagnosticArgument DiagArgs[] = {
+        DiagnosticArgument(0), std::move(VArgs)...
+      };
+      Args.append(DiagArgs + 1, DiagArgs + 1 + sizeof...(VArgs));
+      return fixItReplaceChars(L, L, fixItStringFor(fixIt.ID), std::move(Args));
+    }
+
+    /// Add an insertion fix-it to the currently-active diagnostic.  The
+    /// text is inserted immediately *after* the token specified.
+    template <typename... ArgTypes>
+    InFlightDiagnostic &
+    fixItInsertAfter(SourceLoc L, StructuredFixIt<ArgTypes...> fixIt,
+                     typename detail::PassArgument<ArgTypes>::type... VArgs) {
+      SmallVector<DiagnosticArgument, 3> Args;
+      DiagnosticArgument DiagArgs[] = {
+        DiagnosticArgument(0), std::move(VArgs)...
+      };
+      Args.append(DiagArgs + 1, DiagArgs + 1 + sizeof...(VArgs));
+      return fixItInsertAfter(L, fixItStringFor(fixIt.ID), std::move(Args));
+    }
+
     /// Add a token-based replacement fix-it to the currently-active
     /// diagnostic.
-    InFlightDiagnostic &fixItReplace(SourceRange R, StringRef Str);
+    InFlightDiagnostic &fixItReplace(SourceRange R, StringRef Str) {
+      return fixItReplace(R, Str, {});
```

Comment:
This isn't safe; "Str" might have percents in it. It'd be better to use `"%0"` as the format string and `{Str}` as the args.

Reply:
This happens again later, so I suggest at the least changing the *name* of the arguments that represent format strings even if you still pass them as StringRef.

Reply:
good catch, that would have been a nasty bug!